pkg/vgmanager/vgmanager_controller.go

```diff
@@ -111,64 +157,72 @@ func (r *VGReconciler) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
-	if err != nil {
-		_ = err
-	}
+	_, found := deviceClassMap[volumeGroup.Name]
 
-	for _, deviceClass := range lvmCluster.Spec.DeviceClasses {
-		// ignore deviceClasses whose LabelSelector doesn't match this node
-		// NodeSelectorTerms.MatchExpressions are ORed
-		node := &corev1.Node{}
-		selectsNode, err := NodeSelectorMatchesNodeLabels(node, deviceClass.NodeSelector)
-		if err != nil {
-			r.Log.Error(err, "failed to match nodeSelector to node labels")
-			continue
-		}
-		if !(selectsNode && ToleratesTaints(deviceClass.Tolerations, node.Spec.Taints)) {
-			continue
+	if len(matchingDevices) == 0 {
+		r.Log.Info("no matching devices", "VGName", volumeGroup.Name)
+		if len(delayedDevices) > 0 {
+			return reconcileAgain, nil
+		} else {
+			if found {
+				// Update the status again just to be safe.
+				if statuserr := r.updateStatus(ctx, status, volumeGroup); statuserr != nil {
+					r.Log.Error(statuserr, "failed to update status", "VGName", volumeGroup.Name)
+					return reconcileAgain, nil
+				}
+
+			}
+			return ctrl.Result{}, nil
 		}
-		_, found := deviceClassMap[deviceClass.Name]
+	}
+
+	// create/extend VG and update lvmd config
+	err = r.addDevicesToVG(volumeGroup.Name, matchingDevices)
+	if err != nil {
+		r.Log.Error(err, "failed to create/extend volume group", "VGName", volumeGroup.Name)
+
 		if !found {
-			lvmdConfig.DeviceClasses = append(lvmdConfig.DeviceClasses, &lvmd.DeviceClass{
-				Name:        deviceClass.Name,
-				VolumeGroup: deviceClass.Name,
-				Default:     true,
-			})
+			status.Status = lvmv1alpha1.VGStatusFailed
+			status.Reason = "VGCreationFailed"
+		} else {
+			status.Status = lvmv1alpha1.VGStatusDegraded
+			status.Reason = "VGExtendFailed"
 		}
-		var matchingDevices []internal.BlockDevice
-		remainingValidDevices, matchingDevices, err = filterMatchingDevices(remainingValidDevices, deviceClass)
-		if err != nil {
-			r.Log.Error(err, "could not filterMatchingDevices")
-			continue
-		}
-		if len(matchingDevices) > 0 {
-			// create/extend VG and update lvmd config
-			err = r.addDevicesToVG(deviceClass.Name, matchingDevices)
-			if err != nil {
-				r.Log.Error(err, "could not prepare volume group", "name", deviceClass.Name)
-				continue
-			}
+		if statuserr := r.updateStatus(ctx, status, volumeGroup); statuserr != nil {
+			r.Log.Error(statuserr, "failed to update status", "VGName", volumeGroup.Name)
 		}
+		return reconcileAgain, err
+	}
+
+	if !found {
+		lvmdConfig.DeviceClasses = append(lvmdConfig.DeviceClasses, &lvmd.DeviceClass{
+			Name:        volumeGroup.Name,
+			VolumeGroup: volumeGroup.Name,
+			Default:     true,
+		})
 	}
 
 	// apply and save lvmconfig
 	// pass config to configChannel only if config has changed
 	if !cmp.Equal(existingLvmdConfig, lvmdConfig) {
 		r.Log.Info("updating lvmd config")
 		out, err := yaml.Marshal(lvmdConfig)
-		if err != nil {
-			return ctrl.Result{}, err
+		if err == nil {
+			err = os.WriteFile(controllers.LvmdConfigFile, out, 0600)
```

Comment:
why not move this line to 218 and remove this conditional?

Reply:
The checks work like this:
if unmarshall fails,-> return error
If unmarshall succeeds -> attempt to write
if write fails -> return error

Notice that there is no else check - so the err check will work for both scenarios.